From 7876dc3f9ba404f4961de066e65a595a5ae224b0 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Fri, 29 Sep 2023 14:31:42 +0000
Subject: [PATCH] Use systemd to handle more of the shutdown

Previously we killed kwin manually, then tried to exit the session.
This is racey as kwin will close all connections, before systemd has
realised that the session is ending this can cause all apps to restart.

Also closing plasma-workspace-wayland.target is ineffective, closing the
graphical-session implicitly brings down the session more explicitly.
It'll close the other targets and services.

BUG: 473824
BUG: 409088
BUG: 450531
BUG: 451939
BUG: 450320
BUG: 434403

Tested by looking at `coredumpctl list` after a logout and seeing it go from 10 entries of silly aborts restarting without a display server to 3, with different unrelated causes.
---
 startkde/plasma-shutdown/shutdown.cpp | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/startkde/plasma-shutdown/shutdown.cpp b/startkde/plasma-shutdown/shutdown.cpp
index f5609dcfb..3432f3fbc 100644
--- a/startkde/plasma-shutdown/shutdown.cpp
+++ b/startkde/plasma-shutdown/shutdown.cpp
@@ -69,10 +69,18 @@ void Shutdown::logoutComplete()
 {
     runShutdownScripts();
 
-    // technically this isn't needed in the systemd managed mode, but it seems harmless for now. Guard if it becomes an issue
-    OrgKdeKWinSessionInterface kwinInterface(QStringLiteral("org.kde.KWin"), QStringLiteral("/Session"), QDBusConnection::sessionBus());
-    QDBusPendingReply<> reply = kwinInterface.quit();
-    reply.waitForFinished();
+    auto msg = QDBusMessage::createMethodCall(QStringLiteral("org.freedesktop.systemd1"),
+                                              QStringLiteral("/org/freedesktop/systemd1"),
+                                              QStringLiteral("org.freedesktop.systemd1.Manager"),
+                                              QStringLiteral("StopUnit"));
+    msg << QStringLiteral("graphical-session.target") << QStringLiteral("fail");
+    QDBusReply<QDBusObjectPath> reply = QDBusConnection::sessionBus().call(msg);
+
+    if (!reply.isValid()) {
+        OrgKdeKWinSessionInterface kwinInterface(QStringLiteral("org.kde.KWin"), QStringLiteral("/Session"), QDBusConnection::sessionBus());
+        QDBusPendingReply<> reply = kwinInterface.quit();
+        reply.waitForFinished();
+    }
 
     if (m_shutdownType == KWorkSpace::ShutdownTypeHalt) {
         SessionBackend::self()->shutdown();
-- 
2.42.0

